core/src/main/scala/org/apache/spark/HeartbeatReceiver.scala

```diff
@@ -205,6 +207,13 @@ private[spark] class HeartbeatReceiver(sc: SparkContext, clock: Clock)
             // Note: we want to get an executor back after expiring this one,
             // so do not simply call `sc.killExecutor` here (SPARK-8119)
             sc.killAndReplaceExecutor(executorId)
+            // In case of the executors which are not gracefully shut down, we should remove
+            // lost executors from CoarseGrainedSchedulerBackend manually here (SPARK-27348)
+            sc.schedulerBackend match {
+              case backend: CoarseGrainedSchedulerBackend =>
+                backend.driverEndpoint.send(RemoveExecutor(executorId, ExecutorKilled))
```

Comment:
shall we also do it when we call `sc.killExecutor`?

Reply:
Yeah, thanks for reminding, for `sc.killExecutor`, it still has the problem(maybe it's hardly happened but still possible, e.g., while we call sc.killExecutor, and with the same time the executor crush).
But maybe we should fix this in another PR and JIRA, or I just change the SPARK-27348 description and fix them together, WDYT?

Reply:
IIUC if executors are killed they would enter this heartbeat lost logic anyway.